File: src/tracing.ts
/** Domain-neutral recording of what happened during one trial. */

import { performance } from "node:perf_hooks";
import { type Json, type JsonObject, Message, Step, Transcript, ValidationError, clone, plain } from "./models.js";

export interface UsageRecord { tokens: number | null; cost: number | null; durationMs?: number; name?: string; metadata?: JsonObject }

/** Every environment writes its steps and messages here; graders read the result. */
export class TraceRecorder {
  readonly steps: Step[] = [];
  readonly messages: Message[] = [];
  private readonly goals: string[] = [];

  /** Run a block under a goal; nested calls become sub-goals on every recorded step. */
  goal<T>(name: string, block: () => T): T {
    this.goals.push(name);
    try {
      const result = block();
      if (result instanceof Promise) {
        return result.finally(() => this.goals.pop()) as unknown as T;
      }
      this.goals.pop();
      return result;
    } catch (error) {
      this.goals.pop();
      throw error;
    }
  }

  private goalMetadata(): { goal: string | null; parentGoal: string | null } {
    return { goal: this.goals.at(-1) ?? null, parentGoal: this.goals.length > 1 ? this.goals[this.goals.length - 2] : null };
  }

  appendMessage(role: string, content: string): void {
    this.messages.push(new Message({ role, content, stepIndex: this.steps.length }));
  }

  /** Record actual provider usage, including unknown cost as null. */
  recordUsage({ tokens, cost, durationMs = 0, name = "model", metadata = {} }: UsageRecord): void {
    if ((tokens !== null && tokens < 0) || (cost !== null && cost < 0) || durationMs < 0) {
      throw new ValidationError("Usage values must be nonnegative");
    }
    this.steps.push(new Step({ index: this.steps.length, kind: "model", name, tokens, cost, durationMs, metadata, ...this.goalMetadata() }));
  }

  /** Record a runtime risk estimate before the first state-changing action. */
  recordSignal(probability: number, name = "risk_signal"): void {
    if (this.steps.some((s) => s.metadata.stateChanging)) {
      throw new ValidationError("Calibration signals must be recorded before actions");
    }
    if (!(probability >= 0 && probability <= 1)) throw new ValidationError("A risk signal must lie in [0, 1]");
    this.steps.push(new Step({ index: this.steps.length, kind: "signal", name, result: { riskSignal: probability }, cost: 0, ...this.goalMetadata() }));
  }

  /** Mark that earlier context was compacted: results before this point are no longer visible to the function. */
  recordCompaction(summary = ""): void {
    this.steps.push(new Step({ index: this.steps.length, kind: "compaction", name: "compaction", result: summary || null, cost: 0,
      metadata: { droppedBefore: this.steps.length }, ...this.goalMetadata() }));
  }

  /** Run a tool through the recorder; schema errors are recorded, never hidden. */
  recordTool<T>(name: string, args: JsonObject, call: () => T, options: { stateChanging?: boolean } = {}): T {
    const started = performance.now();
    const previous = this.steps.filter((s) => s.kind === "tool" && s.name === name).at(-1);
    const retry = Boolean(previous && previous.result && typeof previous.result === "object" && !Array.isArray(previous.result) && previous.result.toolFault);
    let schemaValid = true;
    let result: unknown;
    try {
      result = call();
    } catch (error) {
      if (error instanceof ValidationError || error instanceof TypeError || error instanceof RangeError) {
        schemaValid = false;
        result = { ok: false, schemaError: (error as Error).message };
      } else {
        throw error;
      }
    }
    this.steps.push(new Step({
      index: this.steps.length, kind: "tool", name, arguments: clone(args), result: plain(result) as Json,
      durationMs: performance.now() - started, tokens: 0, cost: 0, retry,
      metadata: { schemaValid, stateChanging: Boolean(options.stateChanging) }, ...this.goalMetadata(),
    }));
    return clone(result) as T;
  }

  transcript(): Transcript {
    return new Transcript({ steps: [...this.steps], messages: [...this.messages] });
  }
}

File: src/signal_sdk/tracing.py
"""Domain-neutral recording of what happened during one trial."""

from __future__ import annotations

from contextlib import contextmanager
from copy import deepcopy
from time import perf_counter
from typing import Any, Callable

from .models import Message, Step, Transcript


class TraceRecorder:
    """Every environment writes its steps and messages here; graders read the result."""

    def __init__(self) -> None:
        self._steps: list[Step] = []
        self._messages: list[Message] = []
        self._goals: list[str] = []

    @contextmanager
    def goal(self, name: str):
        """Attach goals and nested sub-goals to recorded execution steps."""
        self._goals.append(name)
        try:
            yield
        finally:
            self._goals.pop()

    def _goal_metadata(self) -> dict[str, Any]:
        return {"goal": self._goals[-1] if self._goals else None,
                "parent_goal": self._goals[-2] if len(self._goals) > 1 else None}

    def append_message(self, role: str, content: str) -> None:
        self._messages.append(Message(role=role, content=content, step_index=len(self._steps)))

    def record_usage(self, *, tokens: int | None, cost: float | None, duration_ms: float = 0.0,
                     name: str = "model", metadata: dict[str, Any] | None = None) -> None:
        """Record actual provider usage, including unknown cost as None."""
        if (tokens is not None and tokens < 0) or (cost is not None and cost < 0) or duration_ms < 0:
            raise ValueError("Usage values must be nonnegative")
        self._steps.append(Step(index=len(self._steps), kind="model", name=name, tokens=tokens,
                                cost=cost, duration_ms=duration_ms, metadata=metadata or {},
                                **self._goal_metadata()))

    def record_signal(self, probability: float, name: str = "risk_signal") -> None:
        """Record a runtime risk estimate before the first state-changing action."""
        if any(s.metadata.get("state_changing") for s in self._steps):
            raise ValueError("Calibration signals must be recorded before actions")
        if not 0 <= probability <= 1:
            raise ValueError("A risk signal must lie in [0, 1]")
        self._steps.append(Step(index=len(self._steps), kind="signal", name=name,
                                result={"risk_signal": probability}, cost=0, **self._goal_metadata()))

    def record_tool(self, name: str, arguments: dict[str, Any], call: Callable[[], Any], *,
                    state_changing: bool = False) -> Any:
        """Run a tool through the recorder; schema errors are recorded, never hidden."""
        started = perf_counter()
        previous = [s for s in self._steps if s.kind == "tool" and s.name == name]
        retry = bool(previous and isinstance(previous[-1].result, dict) and previous[-1].result.get("tool_fault"))
        schema_valid = True
        try:
            result = call()
        except (KeyError, TypeError, ValueError, ArithmeticError) as exc:
            schema_valid = False
            result = {"ok": False, "schema_error": str(exc)}
        self._steps.append(Step(index=len(self._steps), kind="tool", name=name,
                                arguments=deepcopy(arguments), result=deepcopy(result),
                                duration_ms=(perf_counter() - started) * 1000, tokens=0, cost=0.0,
                                retry=retry, metadata={"schema_valid": schema_valid, "state_changing": state_changing},
                                **self._goal_metadata()))
        return deepcopy(result)

    def transcript(self) -> Transcript:
        return Transcript(steps=tuple(self._steps), messages=tuple(self._messages))
